rocksdb_transaction  "const u8" to const i8

compile on Raspberry Pi 5 64


error[E0308]: arguments to this function are incorrect
    --> /home/savages/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/polodb_core-5.1.4/db/rocksdb_transaction.rs:131:13
     |
131  |             ffi::rocksdb_transaction_put(
     |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
     |
note: expected `*const u8`, found `*const i8`
    --> /home/savages/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/polodb_core-5.1.4/db/rocksdb_transaction.rs:133:17
     |
133  |                 key.as_ptr() as *const i8,
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: expected `*const u8`, found `*const i8`
    --> /home/savages/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/polodb_core-5.1.4/db/rocksdb_transaction.rs:135:17
     |
135  |                 value.as_ptr() as *const i8,
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /home/savages/hydroserver/target/debug/build/polodb-librocksdb-sys-f9703e057d043367/out/bindings.rs:5139:12
     |
5139 |     pub fn rocksdb_transaction_put(
     |            ^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
    --> /home/savages/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/polodb_core-5.1.4/db/rocksdb_transaction.rs:152:17
     |
149  |             let value = ffi::rocksdb_transaction_get(
     |                         ---------------------------- arguments to this function are incorrect
...
152  |                 key.as_ptr() as *const i8,
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /home/savages/hydroserver/target/debug/build/polodb-librocksdb-sys-f9703e057d043367/out/bindings.rs:4940:12
     |
4940 |     pub fn rocksdb_transaction_get(
     |            ^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
    --> /home/savages/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/polodb_core-5.1.4/db/rocksdb_transaction.rs:175:17
     |
173  |             ffi::rocksdb_transaction_delete(
     |             ------------------------------- arguments to this function are incorrect
174  |                 self.inner,
175  |                 key.as_ptr() as *const i8,
     |                 ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /home/savages/hydroserver/target/debug/build/polodb-librocksdb-sys-f9703e057d043367/out/bindings.rs:5235:12
     |
5235 |     pub fn rocksdb_transaction_delete(
     |            ^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0308]: mismatched types
    --> /home/savages/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/polodb_core-5.1.4/db/rocksdb_iterator.rs:106:48
     |
106  | ...   ffi::rocksdb_iter_seek(self.inner, key.as_ptr() as *const i8, ke...
     |       ----------------------             ^^^^^^^^^^^^^^^^^^^^^^^^^ expected `*const u8`, found `*const i8`
     |       |
     |       arguments to this function are incorrect
     |
     = note: expected raw pointer `*const u8`
                found raw pointer `*const i8`
note: function defined here
    --> /home/savages/hydroserver/target/debug/build/polodb-librocksdb-sys-f9703e057d043367/out/bindings.rs:1488:12
     |
1488 |     pub fn rocksdb_iter_seek(arg1: *mut rocksdb_iterator_t, k: *const ...
     |            ^^^^^^^^^^^^^^^^^

For more information about this error, try `rustc --explain E0308`.
error: could not compile `polodb_core` (lib) due to 4 previous errors

--- a/.github/workflows/rust.yml
+++ b/.github/workflows/rust.yml
@@ -28,6 +28,8 @@ jobs:
         include:
           - name: Ubuntu
             os: ubuntu-latest
+          - name: Ubuntu ARM64
+            os: ubuntu-24.04-arm
           - name: macOS
             os: macos-latest
           - name: Windows
@@ -46,7 +48,7 @@ jobs:
             ~/.cargo/registry/cache/
             ~/.cargo/git/db/
             target/
-          key: ${{ runner.os }}-cargo-${{ hashFiles('**/Cargo.lock') }}
+          key: ${{ runner.os }}-${{ runner.arch }}-cargo-${{ hashFiles('**/Cargo.lock') }}
 
       - name: Run Rust tests
         run: cargo test --locked --release --verbose --workspace --exclude py-binding-polodb

--- a/src/polodb_core/db/rocksdb_iterator.rs
+++ b/src/polodb_core/db/rocksdb_iterator.rs
@@ -103,7 +103,7 @@ impl RocksDBIteratorInner {
 
     pub fn seek(&self, key: &[u8]) {
         unsafe {
-            ffi::rocksdb_iter_seek(self.inner, key.as_ptr() as *const i8, key.len());
+            ffi::rocksdb_iter_seek(self.inner, key.as_ptr().cast::<c_char>(), key.len());
         }
     }
 

--- a/src/polodb_core/db/rocksdb_transaction.rs
+++ b/src/polodb_core/db/rocksdb_transaction.rs
@@ -130,9 +130,9 @@ impl RocksDBTransactionInner {
 
             ffi::rocksdb_transaction_put(
                 self.inner,
-                key.as_ptr() as *const i8,
+                key.as_ptr().cast::<c_char>(),
                 key.len(),
-                value.as_ptr() as *const i8,
+                value.as_ptr().cast::<c_char>(),
                 value.len(),
                 &mut err,
             );
@@ -149,7 +149,7 @@ impl RocksDBTransactionInner {
             let value = ffi::rocksdb_transaction_get(
                 self.inner,
                 self.read_options.get(),
-                key.as_ptr() as *const i8,
+                key.as_ptr().cast::<c_char>(),
                 key.len(),
                 &mut value_len,
                 &mut err,
@@ -172,7 +172,7 @@ impl RocksDBTransactionInner {
 
             ffi::rocksdb_transaction_delete(
                 self.inner,
-                key.as_ptr() as *const i8,
+                key.as_ptr().cast::<c_char>(),
                 key.len(),
                 &mut err,
             );

